### DISCUSSION TITLE
FastAPI Background task in middleware

### DISCUSSION THREAD

**Discussion by redigaffi** [fastapi/fastapi#4947] [Questions] [answered]
I'm trying to add background tasks to my middleware, but haven't found any official way to do it in their docs, this is what I have tried so far:

```
async def task1():
    logging.info("Waiting ...")
    time.sleep(5)
    logging.info("Waited.")

@app.middleware("http")
async def add_process_time_header(request, call_next, bt=Depends(BackgroundTasks)):
    a = bt.dependency()
    a.add_task(task1)
    a()

    return await call_next(request)
```
This way of injecting bg tasks seems odd as its far from  the examples, for e.g, I have to instantiate BackgroundTasks myself and then call __call__, which the e.g it's not done.

Also, this is blocking my requests, also shouldn't I be able to call async functions without await? I want to completely ignore the results of the background tasks, I dont need results from bg tasks

**Comment by odiseo0** [comment]
You could use a Celery, Redis, RabbitMQ for example and this may help you https://stackoverflow.com/questions/59850517/how-to-run-background-tasks-in-python

**Comment by Kludex** [comment, marked as answer]
You cannot use FastAPI's dependency injection system on middlewares.

Besides, you cannot use `BackgroundTasks` with `BaseHTTPMiddleware`. You can see the limitation [here](https://www.starlette.io/middleware/#basehttpmiddleware).
